x/group/keeper/tally.go

```diff
@@ -0,0 +1,60 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/x/group"
+	"github.com/cosmos/cosmos-sdk/x/group/errors"
+	"github.com/cosmos/cosmos-sdk/x/group/internal/orm"
+)
+
+// Tally is a function that tallies a proposal by iterating through its votes,
+// and returns the tally result without modifying the proposal or any state.
+func (q Keeper) Tally(ctx sdk.Context, p group.Proposal, groupId uint64) (group.TallyResult, error) {
+	// If proposal has already been tallied and updated, then its status is
+	// closed, in which case we just return the previously stored result.
+	if p.Status == group.PROPOSAL_STATUS_CLOSED {
+		return p.FinalTallyResult, nil
+	}
+
+	it, err := q.voteByProposalIndex.Get(ctx.KVStore(q.key), p.Id)
+	if err != nil {
+		return group.TallyResult{}, err
+	}
+	defer it.Close()
+
+	tallyResult := group.DefaultTallyResult()
+
+	var vote group.Vote
+	for {
```

Comment:
Can we not use the standard iterator syntax here?

Reply:
I don't think so, groups is using its internal ORM syntax.

I also don't like the error checking syntax for breaking the loop. Hopefully the new ORM can be made to have a better DX.